src/Traversals/depthfirst.jl

```diff
@@ -146,3 +146,123 @@ end
 end
 
 @traitfn is_cyclic(g::::(!IsDirected), alg::DepthFirst=DepthFirst()) = ne(g) > 0
+
+
+
+mutable struct BiconnectionState{E <: AbstractEdge} <: AbstractTraversalState
+
+    low::Vector{Int}
+    depth::Vector{Int}
+    stack::Vector{E}
+    components::Vector{Vector{E}}
+    id::Int
+    up::Bool
+    lastnode::Int
+    lastlow::Int
+    have_children::Vector{Bool}
+    parent::Vector{Int}
```

Comment:
Since it looks like you only need to track one parent at a time, why not make this just a `eltype(E)` and modify it as necessary during the visit functions?

If you need to keep track of multiples, why not push it on the `stack`?

Reply:
i treat it as a stack, the other stack is for the edges, we need to push and pop those stacks at different times, so we can’t combine them

Reply:
Why do you need to keep track of all parents at all times?

Reply:
I don’t keep track of all the parents, i keep track for the node that we once visited, but haven’t completed yet, which means we will go to them in the future, i e, the ancestor of the current node, the original algorithm save the parent as a local variable, every time we call a function we push its local variables to the stack, and we pop it when the function completes, the recursion can cause multiple function local data to be in the stack at the same time, that what happens here.

Reply:
Can you test something for me? Before you push or pop to `parent`, print out its size.

Reply:
it differ very much from a graph to another, but in general, the size of the parent stack equals when gets poped equals the size of the stack when gets pushed equals the number of the node ancestors in the dfs tree, so if the graph is like star graph the size of `parent` won’t be more than 2 , because no node will ever have more than two ancestors, where ever we star the search, but if the graph is a simple cycle the size of the stack gets pushed in will be [1 2 ... n-1]

Reply:
Can you show me on a simple graph?

Reply:
you mean like the mean, or all the sizes ?

Reply:
Before the `push!` in `visitfn!` and before the `pop!` in `postvisitfn!`, just `println("size of parent = $(length(parent))")`.

Reply:
sorry for taking so long 
```
julia> g = SimpleGraph(10, 30)
{10, 30} undirected simple Int32 graph

julia> biconnected_components2(g)
size of parent = 1
size of parent = 2
size of parent = 3
size of parent = 4
size of parent = 5
size of parent = 6
size of parent = 7
size of parent = 8
size of parent = 9
size of parent = 10
size of parent = 9
size of parent = 8
size of parent = 7
size of parent = 6
size of parent = 5
size of parent = 4
size of parent = 3
size of parent = 2
size of parent = 1
```

Reply:
Ah, I see now. OK. Thanks. So the size of `parent` will never exceed the size of the largest connected component?

Reply:
yes